src/bio/smd/smd_pool.c

```diff
@@ -20,267 +20,317 @@
+	} else if (rc == -DER_NONEXIST) {
+		entry.spe_tgts[0] = tgt_id;
+		entry.spe_blobs[0] = blob_id;
+		entry.spe_tgt_cnt = 1;
+	} else {
+		D_ERROR("Fetch pool "DF_UUID" failed. %d\n",
+			DP_UUID(&key_pool.uuid), rc);
+		goto out;
+	}
 
-	return BTR_CMP_EQ;
+	rc = dbtree_update(smd_store.ss_pool_hdl, &key, &val);
+	if (rc) {
+		D_ERROR("Update pool "DF_UUID" failed. %d\n",
+			DP_UUID(&key_pool.uuid), rc);
+		goto out;
+	}
+out:
+	D_RWLOCK_UNLOCK(&smd_store.ss_rwlock);
+	return rc;
 }
 
-
-static int
-ptab_df_rec_free(struct btr_instance *tins, struct btr_record *rec, void *args)
+int
+smd_pool_unassign(uuid_t pool_id, int tgt_id)
 {
-	struct umem_instance		*umm = &tins->ti_umm;
-
-	if (UMOFF_IS_NULL(rec->rec_off))
-		return -DER_NONEXIST;
+	struct smd_pool_entry	 entry = { 0 };
+	struct d_uuid		 key_pool;
+	d_iov_t			 key, val;
+	int			*tgt_src, *tgt_tgt;
+	uint64_t		*blob_src, *blob_tgt;
+	int			 tgt_idx, rc, move_cnt;
+
+	D_ASSERT(!daos_handle_is_inval(smd_store.ss_pool_hdl));
+
+	uuid_copy(key_pool.uuid, pool_id);
+	D_RWLOCK_WRLOCK(&smd_store.ss_rwlock);
+
+	d_iov_set(&key, &key_pool, sizeof(key_pool));
+	d_iov_set(&val, &entry, sizeof(entry));
+	rc = dbtree_fetch(smd_store.ss_pool_hdl, BTR_PROBE_EQ,
+			  DAOS_INTENT_DEFAULT, &key, NULL, &val);
+	if (rc) {
+		D_ERROR("Fetch pool "DF_UUID" failed. %d\n",
+			DP_UUID(key_pool.uuid), rc);
+		goto out;
+	}
 
-	return umem_free(umm, rec->rec_off);
-}
+	tgt_idx = get_tgt_idx(&entry, tgt_id);
+	if (tgt_idx == -1) {
+		D_ERROR("Pool "DF_UUID" target %d not found.\n",
+			DP_UUID(key_pool.uuid), tgt_id);
+		rc = -DER_NONEXIST;
+		goto out;
+	}
 
-static int
-ptab_df_rec_alloc(struct btr_instance *tins, d_iov_t *key_iov,
-		  d_iov_t *val_iov, struct btr_record *rec)
-{
-	umem_off_t			 npool_off;
-	struct smd_nvme_pool_df		*npool_df;
-	struct pool_tab_key		*pkey = NULL;
-
-	D_ASSERT(key_iov->iov_len == sizeof(struct pool_tab_key));
-	pkey = (struct pool_tab_key *)key_iov->iov_buf;
-
-	npool_off = umem_zalloc(&tins->ti_umm, sizeof(struct smd_nvme_pool_df));
-	if (UMOFF_IS_NULL(npool_off))
-		return -DER_NOSPACE;
-	npool_df = umem_off2ptr(&tins->ti_umm, npool_off);
-	uuid_copy(npool_df->np_info.npi_pool_uuid, pkey->ptk_pid);
-	npool_df->np_info.npi_stream_id = pkey->ptk_sid;
-	memcpy(npool_df, val_iov->iov_buf, sizeof(struct smd_nvme_pool_df));
-	rec->rec_off = npool_off;
-
-	return 0;
-}
+	tgt_src = &entry.spe_tgts[tgt_idx + 1];
+	tgt_tgt = &entry.spe_tgts[tgt_idx];
+	blob_src = &entry.spe_blobs[tgt_idx + 1];
+	blob_tgt = &entry.spe_blobs[tgt_idx];
 
-static int
-ptab_df_rec_fetch(struct btr_instance *tins, struct btr_record *rec,
-		  d_iov_t *key_iov, d_iov_t *val_iov)
-{
-	struct smd_nvme_pool_df		*npool_df;
+	D_ASSERT(entry.spe_tgt_cnt >= (tgt_idx + 1));
+	move_cnt = entry.spe_tgt_cnt - 1 - tgt_idx;
+	entry.spe_tgt_cnt -= 1;
 
-	npool_df = umem_off2ptr(&tins->ti_umm, rec->rec_off);
-	memcpy(val_iov->iov_buf, npool_df, sizeof(struct smd_nvme_pool_df));
+	if (move_cnt && entry.spe_tgt_cnt) {
```

Comment:
Using array works, but this looks a bit clunky IMHO, may be using PMEMOBJ_LISTS would have been a better solution?

Reply:
On the opposite, using list looks clunky to me, because we'll never randomly access an item by it's pointer (item is always located by target ID), so list won't do any better than array here but only introduced some overhead of next/prev pointer and small allocations.